File: scripts/report.js
#!/usr/bin/env node
/**
 * Compact summary of results/report.json (+ cart-results.json) for a live run.
 * Usage: npm run report  [-- --full]
 */
const fs = require('fs');
const path = require('path');

const ROOT = path.join(__dirname, '..');
const resultsDir = path.join(ROOT, 'results');
const full = process.argv.includes('--full');

function readJson(file) {
  try { return JSON.parse(fs.readFileSync(path.join(resultsDir, file), 'utf8')); } catch (_) { return null; }
}

function line(char = '-') { return char.repeat(78); }
function head(text) { console.log(`\n${line('=')}\n${text}\n${line('=')}`); }
function kv(label, value) { console.log(`  ${String(label).padEnd(26)} ${value === undefined || value === null || value === '' ? '—' : value}`); }
function truncate(text, n = 110) {
  const s = String(text === undefined || text === null ? '' : text).replace(/\s+/g, ' ').trim();
  return s.length > n ? `${s.slice(0, n - 1)}…` : s;
}

const report = readJson('report.json');
const cart = readJson('cart-results.json');

if (!report) {
  console.error(`No results/report.json found in ${resultsDir}. Run "npm test" first.`);
  process.exit(1);
}

head(`Fandiem happy flow — ${report.status || 'UNKNOWN'}`);
kv('Title', report.sweepTitle);
kv('Started', report.startedAt);
kv('Finished', report.finishedAt);
kv('Public URL', report.publicUrl);
if (report.error) kv('Error', report.error);

head('Steps');
for (const [i, s] of (report.steps || []).entries()) {
  const status = s.status === 'PASS' ? 'PASS' : 'FAIL';
  console.log(`  ${String(i + 1).padStart(2, '0')}. [${status}] ${s.name} (${Math.round((s.durationMs || 0) / 1000)}s)`);
  if (s.error) console.log(`      error: ${truncate(s.error, 400)}`);
}

if (report.selections) {
  head('Partners');
  kv('Talent', report.selections.talent && report.selections.talent.text);
  kv('Charity', report.selections.charity && report.selections.charity.text);
  kv('Quote title', report.selections.artistQuoteTitle);
  kv('Quote', truncate(report.selections.artistQuote, 90));
  kv('Charity subtitle', truncate(report.selections.charitySubtitle, 90));
}

if (report.media) {
  head('Media');
  kv('Cover', report.media.cover && `${(report.media.cover.files || []).join(',')} via ${report.media.cover.strategy || report.media.cover.error}`);
  kv('Gallery order', (report.media.galleryOrder || []).join(' | '));
  for (const s of report.media.strategies || []) kv(`  ${s.slot}${s.file ? `:${s.file}` : ''}`, s.strategy || s.error);
  kv('Bonus image', report.media.bonus && `${report.media.bonus.file} via ${report.media.bonus.strategy}`);
  for (const c of report.media.typeCoverage || []) kv(`  type ${c.file}`, c.ok ? `ok (${c.strategy})` : `FAILED: ${truncate(c.error, 80)}`);
}

if (report.promotionTabs) {
  head('Promotion Tabs');
  for (const t of report.promotionTabs) {
    kv(`#${t.index} ${truncate(t.title, 40)}`, `${t.descriptionControl ? t.descriptionControl.tag : '?'} | required: ${(t.fieldsRequired || []).join(',') || 'none'} | rawHtml: ${t.rawHtml ? 'on' : 'off'}`);
  }
}
if (report.prizeDetail) {
  head('Prize Detail');
  kv('Emoji', `${report.prizeDetail.emoji} (maxlength ${report.prizeDetail.emojiMaxLength})`);
  kv('Description', truncate(report.prizeDetail.description, 90));
  kv('Modal heading', report.prizeDetail.modalHeading);
}
if (report.entryTiers) {
  head('Entry Tiers');
  kv('Seeded rows', `${report.entryTiers.seededCount} (locked: ${(report.entryTiers.seeded || []).filter(t => t.locked || t.disabled).length})`);
  for (const t of report.entryTiers.seeded || []) {
    kv(`  tier ${t.idx}`, `${t.badge ? `${t.badge} ` : ''}$${t.price} -> ${t.entries} entries ${t.locked ? '(locked)' : ''}`);
  }
  const a = report.entryTiers.added;
  kv(`ADDED tier ${a.idx}`, `$${a.price} -> ${a.entries} entries | ${truncate(a.impact, 60)}`);
}
if (report.bonus) {
  head('Bonus');
  kv('Title', report.bonus.title);
  kv('Image strategy', report.bonus.imageStrategy);
  kv('Entry-tier link', report.bonus.entryTierLink ? report.bonus.entryTierLink.text : '(not linked)');
  kv('Required labels', (report.bonus.labels || []).filter(l => l.required).map(l => l.text).join(', ') || 'none');
}
if (report.sweepsInfo) {
  head('Sweeps Info');
  for (const [k, v] of Object.entries(report.sweepsInfo.values || {})) kv(k.replace(/^.*\[name="|"\]$/g, ''), truncate(v, 70));
  const sw = report.sweepsInfo.switchesAfter || [];
  kv('Switches', sw.map(s => `${s.label}=${s.checked ? 'on' : 'off'}`).join(' | '));
}
if (report.review) {
  head('Review page');
  kv('Attention banner', report.review.attentionBanner ? `YES (${report.review.stepsNeedingAttention} step(s))` : 'no');
  for (const s of report.review.sections || []) {
    const flags = [s.empty ? 'EMPTY' : '', s.needsAttention ? 'NEEDS ATTENTION' : ''].filter(Boolean).join(' ');
    console.log(`  - ${s.section}${flags ? ` [${flags}]` : ''}`);
    if (full) for (const r of s.rows || []) console.log(`      ${r.label}: ${truncate(r.value, 90)}${r.images && r.images.length ? ` (${r.images.length} images)` : ''}`);
  }
  const bad = (report.reviewChecks || []).filter(c => !c.present);
  kv('Value checks', `${(report.reviewChecks || []).length} checked, ${bad.length} missing`);
  for (const b of bad) console.log(`      missing: ${b.section} / ${b.label} -> ${truncate(b.expected, 70)}`);
}
if (report.createNetwork) {
  head('Create Sweep — network');
  kv('URL after submit', report.createNetwork.url);
  kv('POSTs observed', (report.createNetwork.posts || []).length);
  for (const p of report.createNetwork.posts || []) console.log(`      ${p}`);
  for (const p of report.createNetwork.sweepPosts || []) console.log(`      [sweeps] ${p}`);
}
if (report.storefrontChecks) {
  head('Storefront checks');
  for (const c of report.storefrontChecks) {
    console.log(`  [${c.present ? 'OK  ' : 'MISS'}] ${c.required ? '(required) ' : '(recorded) '}${c.label}: ${truncate(c.expected, 70)}`);
  }
  const sn = report.storefront && report.storefront.snapshot;
  if (sn) {
    kv('Title', sn.title);
    kv('ENDS', sn.ends);
    kv('Entries', (sn.entries || []).map(e => `${e.entries} ${e.price} v${e.variant}`).join(' | '));
  }
}
if (report.parity) {
  head('Admin <-> storefront parity');
  kv('Review values missing', (report.parity.reviewMissing || []).length);
  for (const m of report.parity.reviewMissing || []) console.log(`      ${m.section} / ${m.label || '-'}: ${truncate(m.expected, 70)}`);
  kv('Storefront values missing', (report.parity.storefrontMissing || []).length);
  for (const m of report.parity.storefrontMissing || []) console.log(`      ${m.label}: ${truncate(m.expected, 70)}`);
}
if (cart) {
  head('Cart buttons');
  console.log(`  ${cart.length} buttons exercised`);
  for (const c of cart) {
    const state = c.ok ? 'OK  ' : 'FAIL';
    console.log(`  [${state}] #${c.index + 1} ${truncate(c.text, 40)} | POST /cart: ${c.postCartObserved ? 'yes' : 'no'} | cart items: ${c.itemCount}`);
    if (!c.ok) for (const s of c.requests || []) console.log(`        seen: ${s}`);
  }
}

head('Result');
console.log(`  ${report.status}${report.error ? ` — ${truncate(report.error, 200)}` : ''}`);
console.log(`  report: ${path.join(resultsDir, 'report.json')}${cart ? `\n  cart:   ${path.join(resultsDir, 'cart-results.json')}` : ''}`);

File: scripts/report.py
#!/usr/bin/env python3
"""
report.py — Python port of report.js
Shows Fandiem happy flow report with issue details: was it visible, what did code do, what error received
Usage: python scripts/report.py [--full]
       py scripts/report.py
       npm run report (still works via node scripts/report.js)
       python -m scripts.report
"""
import json
import pathlib
import sys

ROOT = pathlib.Path(__file__).resolve().parent.parent
resultsDir = ROOT / "results"
full = "--full" in sys.argv

def read_json(name):
    try:
        return json.loads((resultsDir / name).read_text(encoding="utf-8"))
    except:
        return None

def line(ch='-'): return ch*78
def head(text): print(f"\n{line('=')}\n{text}\n{line('=')}")
def kv(label, value):
    v = value if value not in (None, "", []) else "—"
    # handle list
    if isinstance(v, list):
        v = ", ".join(str(x) for x in v) if v else "—"
    print(f"  {str(label).ljust(26)} {v}")
def truncate(text, n=110):
    s = str(text if text is not None else "").replace("\n"," ").replace("\r"," ").strip()
    s = " ".join(s.split())
    return s[:n-1]+"…" if len(s)>n else s

report = read_json("report.json")
cart = read_json("cart-results.json")

if not report:
    print(f"No results/report.json found in {resultsDir}. Run \"py scripts/run.py\" first.", file=sys.stderr)
    sys.exit(1)

head(f"Fandiem happy flow — {report.get('status','UNKNOWN')}")
kv("Title", report.get("sweepTitle"))
kv("Started", report.get("startedAt"))
kv("Finished", report.get("finishedAt"))
kv("Public URL", report.get("publicUrl"))
if report.get("error"):
    kv("Error", report.get("error"))
    # Show issue clearly as requested
    print("\n  Issue detail:")
    print(f"    Visible on page: {report.get('error','')[:500]}")
    # Try to show media debug if available
    if report.get("media"):
        print(f"    Media debug: {json.dumps(report['media'], indent=6)[:2000]}")

head("Steps")
for i, s in enumerate(report.get("steps", []), 1):
    status = "PASS" if s.get("status")=="PASS" else "FAIL"
    print(f"  {i:02d}. [{status}] {s.get('name')} ({round((s.get('durationMs') or 0)/1000)}s)")
    if s.get("error"):
        print(f"      error: {truncate(s['error'], 400)}")
        # State clearly what code did
        if "Gallery" in s.get("name","") or "Campaign" in s.get("name",""):
            print(f"      what code did: tried reveal_file_inputs, list_file_inputs, setInputFiles on div.grid input[multiple] and button+input sibling, drop on button, click+upload")
            print(f"      was it visible: check report.media.fileInputs and report.media.galleryOrder")
            if report.get("media"):
                print(f"      fileInputs: {report['media'].get('fileInputs')}")
                print(f"      strategies: {report['media'].get('strategies')}")

if report.get("selections"):
    head("Partners")
    sel = report["selections"]
    kv("Talent", sel.get("talent",{}).get("text") if isinstance(sel.get("talent"), dict) else sel.get("talent"))
    kv("Charity", sel.get("charity",{}).get("text") if isinstance(sel.get("charity"), dict) else sel.get("charity"))
    kv("Quote title", sel.get("artistQuoteTitle"))
    kv("Quote", truncate(sel.get("artistQuote"), 90))
    kv("Charity subtitle", truncate(sel.get("charitySubtitle"), 90))

if report.get("media"):
    head("Media")
    m = report["media"]
    cover = m.get("cover")
    if cover:
        kv("Cover", f"{','.join(cover.get('files') or [])} via {cover.get('strategy') or cover.get('error')}")
        if cover.get("error"):
            print(f"      Cover error detail: {cover['error']}")
    kv("Gallery order", " | ".join(m.get("galleryOrder") or []))
    for s in m.get("strategies") or []:
        kv(f"  {s.get('slot')}{':'+s.get('file') if s.get('file') else ''}", s.get("strategy") or s.get("error"))
        if s.get("error"):
            print(f"        strategy error: {s['error'][:400]}")
    bonus = m.get("bonus")
    if bonus:
        kv("Bonus image", f"{bonus.get('file')} via {bonus.get('strategy')}")
    for c in m.get("typeCoverage") or []:
        kv(f"  type {c.get('file')}", f"ok ({c.get('strategy')})" if c.get("ok") else f"FAILED: {truncate(c.get('error'),80)}")
    # Visibility check as requested
    kv("Items text", m.get("itemsText"))
    kv("FileInputs at start", str(m.get("fileInputs")))
    if not m.get("galleryOrder"):
        print("      Issue: Gallery is required but galleryOrder is empty - upload did not create tiles (0 items)")
        print("      Code tried: reveal (make visible), setInputFiles on div.grid input[multiple] with 3-5 files batch, fallback one-by-one, drop, click+upload")
        print("      Visible check: fileInputs was empty means list_file_inputs returned [] even though reveal said 2 inputs - now fixed to handle quoted JSON")

if report.get("promotionTabs"):
    head("Promotion Tabs")
    for t in report["promotionTabs"]:
        kv(f"#{t.get('index')} {truncate(t.get('title'),40)}", f"{(t.get('descriptionControl') or {}).get('tag','?')} | required: {','.join(t.get('fieldsRequired') or []) or 'none'}")

if report.get("prizeDetail"):
    head("Prize Detail")
    kv("Emoji", f"{report['prizeDetail'].get('emoji')} (maxlength {report['prizeDetail'].get('emojiMaxLength')})")
    kv("Description", truncate(report['prizeDetail'].get('description'),90))

if report.get("entryTiers"):
    head("Entry Tiers")
    kv("Seeded rows", f"{report['entryTiers'].get('seededCount')}")
    a = report['entryTiers'].get('added')
    if a:
        kv(f"ADDED tier {a.get('idx')}", f"${a.get('price')} -> {a.get('entries')} entries | {truncate(a.get('impact'),60)}")

if report.get("bonus"):
    head("Bonus")
    kv("Title", report['bonus'].get('title'))

if report.get("sweepsInfo"):
    head("Sweeps Info")
    for k,v in (report['sweepsInfo'].get('values') or {}).items():
        kv(k.replace('input[name="','').replace('"]',''), truncate(v,70))

if report.get("review"):
    head("Review page")
    kv("Attention banner", f"YES ({report['review'].get('stepsNeedingAttention')} step(s))" if report['review'].get('attentionBanner') else "no")

if report.get("storefrontChecks"):
    head("Storefront checks")
    for c in report["storefrontChecks"]:
        print(f"  [{'OK  ' if c.get('present') else 'MISS'}] {'(required) ' if c.get('required') else '(recorded) '}{c.get('label')}: {truncate(c.get('expected'),70)}")

head("Result")
print(f"  {report.get('status')}{(' — '+truncate(report.get('error'),200)) if report.get('error') else ''}")
print(f"  report: {resultsDir / 'report.json'}")
if cart:
    print(f"  cart:   {resultsDir / 'cart-results.json'}")
    head("Cart buttons")
    print(f"  {len(cart)} buttons")
